nil-discussion.md

```diff
@@ -0,0 +1,180 @@
+   - `property :x, Integer, default: "10"` is INVALID.
+   - `property :x, Integer, default: lazy { "10" }` is INVALID.
+
+3. Properties must never return invalid values.
+
+   This is a general rule: we're trying to make properties that don't return
+   unexpected values (this is why coercion is so powerful, for example). This
+   rule can be satisfied by making `nil` an expected value, but it *still* needs
+   to be spelled out.
+
+4. Anything you can pass to a setter, you can put in a default.
+
+   Corollary: default values (and lazy default values) are coerced.
+
+   - `property :x, Integer, coerce: proc { |v| v.to_i }, default: "10"` is VALID.
+   - `property :x, Integer, coerce: proc { |v| v.to_i }, default: lazy { "10" }` is VALID.
+
+5. Copying a property value from A to B always works.
+
+   Corollary, if a getter can successfully return a value, then the setter must
+   allow it to be set.
+
+   Combined with the above, this means:
+
+   - `x(nil)` with `property :x, String` is VALID.
+   - `x(nil)` with `property :x` is VALID.
+
+6. Properties with valid non-nil default values should not accept `nil`.
+
+   - `x(nil)` with `property :x, Symbol, default: :foo` is INVALID.
+
+7. Lazy values must validate the same as non-lazy values.
+
+   Particularly, this means we can't say `nil` is valid for `default: nil`
+   but *not* valid for `default: lazy { nil }` (or more likely,
+   `default: lazy { other_property_that_could_be_nil }`).
+
+## Possible Solutions
+
+The possible solutions I can come up with are below. I think the one that meets
+all our criteria with the least complexity is "Properties with no default or
+`default: nil` are valid", with "Properties with no default are valid" running
+a close second for increased parsimony.
+
+### `nil` is always validated
+
+`nil` will fail validation if you don't allow it. This passes all our rules
+except #1, which means this common construction would simply be invalid.
+
+```
+# This would throw ValidationError when you define the property
+property :x, String
+```
+
+### `nil` is always valid
+
+By declaring `nil` as always valid, we violate the spirit but not the letter of #2,
+and allow `property :x, String` to work. However, we completely violate #5,
+making `nil` unnecessarily valid for this property:
+
+```
+property :x, Symbol, default: :foo
+```
+
+Or for this more complex but still important case:
+
+```
+property :path, String, default: "/opt/chefdk"
+property :config_path, String, default: lazy { "#{path}/default_config.ini" }
+```
+
+### Properties with no default treat `nil` as valid
+
+This is tied with the next one for my favorite.
+
+This passes all of the rules, and is simple. If the default is not specified,
+then we are giving you a default, and we should take our own steps to make sure
+it is a valid one.
+
+The only issue is that these two would NOT be equivalent:
+
+```
+property :x, String
+property :x, String, default: nil # ValidationError when declaring this
```

Comment:
This being a ValidationError is OK with me in theory but is exactly the kind of minor annoyance that leads to frustration I think.  "If the 'default default' is nil, why can't I set it to nil?", my imagined user says.